refactor(linter): `exhaustive-deps`: remove usage of `Atom`

camchenry opened the pull request:
We were calling `.as_str()` everywhere in this file already, and we didn't need the `Atom` type specifically, it just happened to be the type for identifier names. This refactor will make changing the identifier type later easier.

camc314 commented on crates/oxc_linter/src/rules/react/exhaustive_deps.rs:
```diff
@@ -1443,10 +1438,10 @@ impl<'a> Visit<'a> for ExhaustiveDepsVisitor<'a, '_> {
                     reference_id,
                     span: ident.span,
-                    chain: vec![prop],
+                    chain: vec![&prop],
```
This is incorrect, else the CI for this PR would be failing.

camchenry committed:
refactor(linter): `exhaustive-deps`: remove usage of `Atom` (#16418)

We were calling `.as_str()` everywhere in this file already, and we didn't need the `Atom` type specifically, it just happened to be the type for identifier names. This refactor will make changing the identifier type later easier.

overlookmotel commented:
@camchenry Anywhere that we do comparisons between identifiers (by which I mean the name of `BindingIdentifier`s or `IdentifierReference`s), we'll want to keep the string as an `Ident` as much as possible, rather than dereferencing to `&str` - because comparing 2 x `Ident`s will be much faster than comparing 2 x `&str`s.

Once you introduce the `Ident` type, we'll need to go through the codebase and make changes in many places. I wonder if that'll be easier to do if we keep `Atom`s for now, as we'll get red squiggles everywhere that needs to change. Whereas if we've moved code over to `&str` (like these PRs do), compiler won't flag that any changes are needed, and the code will work, but be less performant than it could be.

Maybe you already have a clear plan in mind already, and so these changes make sense. But could we possibly discuss a bit more on https://github.com/oxc-project/backlog/issues/193?

overlookmotel commented:
I've added some thoughts on https://github.com/oxc-project/backlog/issues/193.

Just to say, I don't want to block your progress. We don't have to implement everything in one go, and maybe your approach in these PRs is the most pragmatic starting point. But I just wanted to ask the question before we start merging.

camchenry commented:
@overlookmotel Sorry for jumping right in here! I hadn't considered the `==` check aspect. I was figuring that `Atom` usage here was mostly incidental, but the original goal was to reduce the size of whatever PR introduces the `Ident` type. However, there might be a different way to accomplish the same thing without compromising on performance. Since the point was to not regress performance though, I'll close this approach. We can discuss more in https://github.com/oxc-project/backlog/issues/193.

overlookmotel commented:
No not at all. *Thank you* for getting into this. I've wanted to tackle it for ages, but I've been tied up. Really glad you're taking the initiative. Sorry I've pissed on your parade.

camchenry commented:
> No not at all. _Thank you_ for getting into this. I've wanted to tackle it for ages, but I've been tied up. Really glad you're taking the initiative. Sorry I've pissed on your parade.

No worries at all. 90% of performance work is just discarding code that doesn't work anyway 😄
